### PR TITLE
Handle symlinking edge cases

### PR DESCRIPTION
**Pull request by withinboredom**
This one is interesting — though I’m not sure the best way to provide a test. I will have to look into maybe an integration test because it is a careful dance between how we resolve paths in the Caddy module vs. workers. I looked into making a proper change (literally using the same logic everywhere), but I think it is best to wait until #1646 is merged.

But anyway, this change deals with some interesting edge cases. I will use gherkin to describe them:

```gherkin
Feature: FrankenPHP symlinked edge cases
  Background: 
    Given a `test` folder
    And a `public` folder linked to `test`
    And a worker script located at `test/index.php`
    And a `test/nested` folder
    And a worker script located at `test/nested/index.php`
  Scenario: neighboring worker script
    Given frankenphp located in the test folder
    When I execute `frankenphp php-server --listen localhost:8080 -w index.php` from `public`
    Then I expect to see the worker script executed successfully
  Scenario: nested worker script
    Given frankenphp located in the test folder
    When I execute `frankenphp --listen localhost:8080 -w nested/index.php` from `public`
    Then I expect to see the worker script executed successfully
  Scenario: outside the symlinked folder
    Given frankenphp located in the root folder
    When I execute `frankenphp --listen localhost:8080 -w public/index.php` from the root folder
    Then I expect to see the worker script executed successfully
  Scenario: specified root directory
    Given frankenphp located in the root folder
    When I execute `frankenphp --listen localhost:8080 -w public/index.php -r public` from the root folder
    Then I expect to see the worker script executed successfully    
```

Trying to write that out in regular English would be more complex IMHO.

These scenarios should all pass now with this PR.

### CODE REVIEW

File: worker.go [hunk 1]
```diff
@@ -74,7 +75,14 @@ func getWorkerKey(name string, filename string) string {
...
 
 func newWorker(o workerOpt) (*worker, error) {
-	absFileName, err := fastabs.FastAbs(o.fileName)
+	//Order is important!
+	//This order ensures that FrankenPHP started from inside a symlinked directory will properly resolve any paths.
+	//If it is started from outside a symlinked directory, it is resolved to the same path that we use in the Caddy module.
+	absFileName, err := filepath.EvalSymlinks(o.fileName)
+	if err != nil {
+		return nil, fmt.Errorf("worker filename is invalid %q: %w", o.fileName, err)
+	}
```
Comment (AlliBalliBaba) [inline worker.go:120]: I like it 👍 explicitly checking if a worker exists is something I actually wanted to do for a long time

File: caddy/module.go [hunk 2]
```diff
@@ -148,7 +148,10 @@ func (f *FrankenPHPModule) ServeHTTP(w http.ResponseWriter, r *http.Request, _ c
 		if documentRoot == "" && frankenphp.EmbeddedAppPath != "" {
 			documentRoot = frankenphp.EmbeddedAppPath
 		}
-		documentRootOption = frankenphp.WithRequestDocumentRoot(documentRoot, *f.ResolveRootSymlink)
+		//If we do not have a resolved document root, then we cannot resolve the symlink of our cwd because it may
+		//resolve to a different directory than the one we are currently in.
+		//This is especially important if there are workers running.
+		documentRootOption = frankenphp.WithRequestDocumentRoot(documentRoot, false)
```
Comment (AlliBalliBaba) [inline caddy/module.go:196]: Not checking for symlinks here seems like it might break some setups. Definitely would be nice to test all the edge cases somehow (should be possible with caddy_tests?). 

Tbh though, I don't know what the current edge cases with symlinks are

File: worker.go (hunk 1 above)
Comment (AlliBalliBaba) [reply worker.go:120]: Or wait, evalSymlinks will just evaluate symlinks. In that case it would also be nice to error right away if there's no file there. Can also be done in another PR though

Comment (withinboredom) [reply worker.go:120]: it errors if the file doesn't exist (I had to update a failing test -- haven't pushed the commit yet)

File: caddy/module.go (hunk 2 above)
Comment (withinboredom) [reply caddy/module.go:196]: I’m pretty sure we don’t want to evaluate symlinks here. In the case where a document root was expressly given (such as prod setups), we will never even enter this branch. In the case where a document root isn’t given, we need to be compatible with how we resolve workers. There is an edge case here that I didn’t cover, and I’m working out how to handle it: chained symlinks.

I suspect I’ll have to go back to the drawing board.

**Comment by henderkes** [comment]
do the edge cases only affect workers, not regular php requests?

File: caddy/module.go [hunk 3]
```diff
@@ -148,7 +148,10 @@ func (f *FrankenPHPModule) ServeHTTP(w http.ResponseWriter, r *http.Request, _ c
 		if documentRoot == "" && frankenphp.EmbeddedAppPath != "" {
 			documentRoot = frankenphp.EmbeddedAppPath
 		}
-		documentRootOption = frankenphp.WithRequestDocumentRoot(documentRoot, *f.ResolveRootSymlink)
+		//If we do not have a resolved document root, then we cannot resolve the symlink of our cwd because it may
+		//resolve to a different directory than the one we are currently in.
+		//This is especially important if there are workers running.
```
Comment (dunglas) [inline caddy/module.go]:
```suggestion
		// If we do not have a resolved document root, then we cannot resolve the symlink of our cwd because it may
		// resolve to a different directory than the one we are currently in.
		// This is especially important if there are workers running.
```

File: worker.go [hunk 4]
```diff
@@ -74,7 +75,14 @@ func getWorkerKey(name string, filename string) string {
 }
 
 func newWorker(o workerOpt) (*worker, error) {
-	absFileName, err := fastabs.FastAbs(o.fileName)
+	//Order is important!
+	//This order ensures that FrankenPHP started from inside a symlinked directory will properly resolve any paths.
+	//If it is started from outside a symlinked directory, it is resolved to the same path that we use in the Caddy module.
```
Comment (dunglas) [inline worker.go]:
```suggestion
	// Order is important!
	// This order ensures that FrankenPHP started from inside a symlinked directory will properly resolve any paths.
	// If it is started from outside a symlinked directory, it is resolved to the same path that we use in the Caddy module.
```

**Review by dunglas — CHANGES_REQUESTED** [review]

**Comment by withinboredom** [comment]
@henderkes yes, only workers. I should have a nice test suite this weekend. Manually testing this is annoying.

**Comment by henderkes** [comment]
@withinboredom should I pick this up or do you have something stashed away to commit?

**Comment by withinboredom** [comment]
TBH, I forgot this existed. If you want to take a gander, go for it!

**Comment by henderkes** [comment]
In adding a test manually after AI generated the bunch, of course the manually generated one errors. Edge case when using relative worker scripts:

```Caddyfile
localhost
php {
    root /path/to/symlink
    worker index.php
}
```

File: caddy/caddy_test.go [hunk 5]
```diff
@@ -1500,3 +1500,269 @@ func TestLog(t *testing.T) {
...
 	)
 }
+
+// TestSymlinkNeighboringWorkerScript tests executing a worker script from within a symlinked directory
+// Scenario: neighboring worker script
+//
+//	Given frankenphp located in the test folder
+//	When I execute `frankenphp php-server --listen localhost:8080 -w index.php` from `public`
+//	Then I expect to see the worker script executed successfully
+func TestSymlinkNeighboringWorkerScript(t *testing.T) {
```
Comment (dunglas) [inline caddy/caddy_test.go]: Couldn't you use [subtests](https://go.dev/blog/subtests) instead a separated tests?

File: testdata/symlinks/public [hunk 6]
```diff
@@ -0,0 +1 @@
+test
```
Comment (dunglas) [inline testdata/symlinks/public:1]:
```suggestion
test

```

File: testdata/symlinks/test/document-root.php [hunk 7]
```diff
@@ -0,0 +1,13 @@
+<?php
+
+if (isset($_SERVER['FRANKENPHP_WORKER']) && $_SERVER['FRANKENPHP_WORKER'] === '1') {
```
Comment (dunglas) [inline testdata/symlinks/test/document-root.php]:
```suggestion
if (isset($_SERVER['FRANKENPHP_WORKER'])) {
```

File: testdata/symlinks/test/index.php [hunk 8]
```diff
@@ -0,0 +1,13 @@
+<?php
+
+if (!isset($_SERVER['FRANKENPHP_WORKER']) || $_SERVER['FRANKENPHP_WORKER'] !== '1') {
```
Comment (dunglas) [inline testdata/symlinks/test/index.php]: This cannot happen:

```suggestion
if (!isset($_SERVER['FRANKENPHP_WORKER'])) {
```

File: testdata/symlinks/test/nested/index.php [hunk 9]
```diff
@@ -0,0 +1,13 @@
+<?php
+
+if (!isset($_SERVER['FRANKENPHP_WORKER']) || $_SERVER['FRANKENPHP_WORKER'] !== '1') {
```
Comment (dunglas) [inline testdata/symlinks/test/nested/index.php]:
```suggestion
if (!isset($_SERVER['FRANKENPHP_WORKER'])) {
```

**Review by dunglas — APPROVED** [review]

File: testdata/symlinks/public (hunk 6 above)
Comment (henderkes) [reply testdata/symlinks/public:1]: This is a symlink, not a text file. I don't think we should be changing the content

File: caddy/caddy_test.go (hunk 5 above)
Comment (henderkes) [reply caddy/caddy_test.go]: I'll refactor it tomorrow, thanks for the link.

Comment (henderkes) [reply caddy/caddy_test.go]: Oh and Happy New Year!

Comment (dunglas) [reply caddy/caddy_test.go]: Happy new year!! And thanks.
